FunKiiU.py

```diff
@@ -1,374 +1,378 @@
+            titlekeys_data = json.load(data_file)
+
+    for title_id in titles:
+        if not check_title_id(title_id):
+            print('The Title ID(s) must be 16 hexadecimal characters long')
+            print('{} - is not ok.'.format(title_id))
+            sys.exit(0)
+        title_key = None
+        name = None
+
+        if keys:
+            title_key = keys.pop()
+        elif onlinekeys or onlinetickets:
+            title_data = next((t for t in titlekeys_data if t['titleID'] == title_id.lower()), None)
+
+            if not title_data:
+                print("ERROR: Could not find key for ID {}, skipping".format(title_id))
+                continue
+
+            elif onlinetickets:
+                if not title_data['ticket']:
+                    print('ERROR: Ticket not available online for {}'.format(title_id))
                     continue
 
+            elif onlinekeys:
+                title_key = title_data['titleKey']
 
+            name = title_data.get('name', None)
+
+        if not (title_key or onlinetickets):
+            print('ERROR: Could not find title or ticket for {}'.format(title_id))
+            continue
 
-            if (titleid in titlelist):
-                #if we want the ticket, but it doesn't exist, skip
-                if arguments.onlinetickets:
-                    if not ticketexists:
-                        print 'ERROR: Cannot find ticket on ' + keysite + '\n'
-                        continue
-                #if we want key, check if it exists + ok, else skip
-                if arguments.onlinekeys:
-                    if key and (len(key) is 32) and all(c in string.hexdigits for c in key):
-                        pass
-                    else:
-                        print 'ERROR: No key/bad key on ' + keysite + '\n'
-                        continue
-                processTitleID(titleid, key)
-
-            if arguments.all:
-                if arguments.onlinetickets:
-                    if not ticketexists:
-                        continue
-                if arguments.onlinekeys:
-                    if key and (len(key) is 32) and all(c in string.hexdigits for c in key):
-                        pass
-                    else:
-                        continue
-                processTitleID(titleid, key, name)
-
-
-else:
-    processTitleID(titlelist[0], arguments.key, '')
\ No newline at end of file
+        process_title_id(title_id, title_key, name, output_dir, retry_count, onlinetickets, patch_demo, patch_dlc)
+
+    if download_all:
+        for title_data in titlekeys_data:
+            title_id = title_data['titleID']
+            title_key = title_data.get('titleKey', None)
+            name = title_data.get('name', None)
+            typecheck = title_id[4:8]
+
+            # skip updates
+            if typecheck in ('000e', '8005', '800f') or int(typecheck, 16) & 0x10:
+                continue
+            elif title_id in titles:
```

Comment:
Why are title ids that we specify checked and ignored when using `download_all`?

It seems if we would do:
`FunKii.py -title pikmin3titleID -all`
that this would download everything apart from mario kart 8.
Which seems strange. I will add a check that ensures there are no title ids specified when using the -all option.